### DISCUSSION TITLE
Replacement for actix-web?

### DISCUSSION THREAD

**Discussion by dbr** [taskchampion/taskchampion#155] [General]
`actix-web` brings in several hundred crates as dependencies, and adds a lot to the build time (over a minute for me). Given the sync server has a nice small API (about two end points serving a custom document format), it probably don't benefit much from a large web-framework - instead we could probably use a lower level HTTP server crate

- The most obvious choice would be `hyper` - not the lightest of weight crates (around 17 sec to build for me), but reasonably constrained in scope. Still 0.x but widely used enough that the minor version bumps are treated much like major version bumps (each minor release seems quite well supported, e.g there was 35 patch releases for v0.12)
- ..there isn't actually that many other decent options I can find. There is `tiny_http` I've used before via roullie, it seems quite stable but not too actively maintained, and thread-pool based. There's also a pile of other experimental web server/frameworks

**Comment by djmitche** [comment]
This sounds like a good idea, but pretty low priority -- that build time is a one-time hit, after all.

I think `hyper` is very low-level, right?  We'd have to write our own routing support, for example.  Is that hit in complexity worth the savings in compile time?

And, what kind of future improvements might we make that Actix would support, while Hyper would render difficult?  If we added endpoints to manage clientIds, for example?  Or authentication middleware?

**Reply by dbr** [reply]
It's not necessarily a one-time hit given CI builds, testing against multiple Rust configurations (e.g minimum supported version + current stable, debug vs release mode), running tests with code-coverage which currently requires building under a nightly compiler

..but I definitely agree it's a low priority thing!

> And, what kind of future improvements might we make that Actix would support, while Hyper would render difficult? If we added endpoints to manage clientIds, for example? Or authentication middleware?

I've not used it directly so I'm not 100% sure, but Hyper is indeed fairly low-level; e.g there is no built-in routing as you suspect. Currently we could just do `if is_get && request_path == "/..." { ... } else if ...` branch, and if there are a few more complex URL's added we could reasonably implement those with a simple regex or two.

Authentication process is an interesting point, although I wonder if using hyper may give us more options (e.g we might be able to use the underlying libraries used by actix-web, or rocket, etc instead of being tied to just actix-specific middleware), but that's is speculation on my part currently as I've not looked into this at all.

**Comment by djmitche** [comment]
Maybe the right approach is #167?
